### PR TITLE
refactor: change wrapper setting

### PR DESCRIPTION
**Pull request by Gaiejj**
## Description

refactor: change wrapper setting

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/PKU-MARL/omnisafe/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://omnisafe.readthedocs.io/en/latest/developer/contributing.html) guide. (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format`. (**required**)
- [x] I have checked the code using `make lint`. (**required**)
- [x] I have ensured `make test` pass. (**required**)

### CODE REVIEW

File: omnisafe/algorithms/off_policy/ddpg.py [hunk 1]
```diff
@@ -13,15 +13,15 @@
 # limitations under the License.
 # ==============================================================================
 """Implementation of the DDPG algorithm."""
-
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/ddpg.py:16]: This blank line is required.

File: omnisafe/algorithms/off_policy/ddpg_lag.py [hunk 2]
```diff
@@ -14,23 +14,28 @@
...
 """Implementation of the Lagrange version of the DDPG algorithm."""
 
+from typing import Dict, NamedTuple, Tuple
+
+import torch
+
 from omnisafe.algorithms import registry
 from omnisafe.algorithms.off_policy.ddpg import DDPG
 from omnisafe.common.lagrange import Lagrange
+from omnisafe.models.constraint_actor_q_critic import ConstraintActorQCritic
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/ddpg_lag.py:26]: duplicate line, one blank line.

File: omnisafe/algorithms/off_policy/sac.py [hunk 3]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 # ==============================================================================
 """Implementation of the SAC algorithm."""
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/sac.py:15]: blank line.

File: omnisafe/algorithms/off_policy/sac_lag.py [hunk 4]
```diff
@@ -13,25 +13,27 @@
 # limitations under the License.
 # ==============================================================================
 """Implementation of the Lagrange version of the SAC algorithm."""
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/sac_lag.py:15]: blank line.

File: omnisafe/algorithms/off_policy/sddpg.py [hunk 5]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 # ==============================================================================
 """Implementation of the SDDPG algorithm."""
+from typing import Dict, NamedTuple, Tuple
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/sddpg.py:16]: blank line.

File: omnisafe/algorithms/off_policy/td3.py [hunk 6]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 # ==============================================================================
 """Implementation of the TD3 algorithm."""
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/td3.py:15]: blank line.

File: omnisafe/algorithms/off_policy/td3_lag.py [hunk 7]
```diff
@@ -13,23 +13,27 @@
 # limitations under the License.
 # ==============================================================================
 """Implementation of the Lagrange version of the TD3 algorithm."""
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/td3_lag.py:15]: blank line.

File: omnisafe/algorithms/on_policy/early_terminated/ppo_early_terminated.py [hunk 8]
```diff
@@ -28,6 +30,11 @@ class PPOEarlyTerminated(PPO):
         URL: https://arxiv.org/abs/2107.04200
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/early_terminated/ppo_early_terminated.py:30]: why is this line different from Reference's comment in off-policy algorithms?

File: omnisafe/algorithms/on_policy/first_order/focops.py [hunk 9]
```diff
@@ -49,125 +57,146 @@ def __init__(self, env_id, cfgs) -> None:
...
         ent = dist.entropy().mean().item()
         pi_info = dict(kl=approx_kl, ent=ent, ratio=ratio.mean().item())
 
         return loss_pi, pi_info
 
-    def update(self):
-        """Update."""
-        raw_data, data = self.buf.pre_process_data()
-        # First update Lagrange multiplier parameter
+    # pylint: disable=too-many-locals
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/first_order/focops.py:120]: should we use disable？

File: omnisafe/algorithms/on_policy/naive_lagrange/ppo_lag.py [hunk 10]
```diff
@@ -14,27 +14,32 @@
...
 from omnisafe.algorithms import registry
-from omnisafe.algorithms.on_policy.base.policy_gradient import PolicyGradient
+from omnisafe.algorithms.on_policy.base.ppo import PPO
 from omnisafe.common.lagrange import Lagrange
 
 
 @registry.register
-class PPOLag(PolicyGradient, Lagrange):
+class PPOLag(PPO, Lagrange):
     """The Lagrange version of the PPO algorithm.
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/naive_lagrange/ppo_lag.py:29]: Do I need to cite the safety gym paper here?

File: omnisafe/algorithms/on_policy/second_order/cpo.py [hunk 11]
```diff
@@ -32,52 +34,73 @@
...
 
+    CPO is a derivative of TRPO.
+
     References:
-        Title: Constrained Policy Optimization
-        Authors: Joshua Achiam, David Held, Aviv Tamar, Pieter Abbeel.
-        URL: https://arxiv.org/abs/1705.10528
+        - Title: Constrained Policy Optimization
+        - Authors: Joshua Achiam, David Held, Aviv Tamar, Pieter Abbeel.
+        - URL: https://arxiv.org/abs/1705.10528
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/second_order/cpo.py:42]: URL style.

File: omnisafe/algorithms/on_policy/second_order/pcpo.py [hunk 12]
```diff
@@ -37,42 +39,82 @@ class PCPO(TRPO):
         URL: https://arxiv.org/abs/2010.03152
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/second_order/pcpo.py:39]: URL styple.

File: omnisafe/algorithms/on_policy/second_order/pcpo.py [hunk 13]
```diff
@@ -119,31 +188,77 @@ def algorithm_specific_logs(self):
...
+            log_p (torch.Tensor): Log probability.
+            cost_adv (torch.Tensor): Cost advantage.
+        """
+        _, _log_p = self.actor_critic.actor(obs, act)
+        ratio = torch.exp(_log_p - log_p)
+        cost_loss = (ratio * cost_adv).mean()
         info = {}
         return cost_loss, info
 
+    # pylint: disable=too-many-locals,invalid-name,too-many-arguments
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/second_order/pcpo.py:221]: disable-next

File: omnisafe/algorithms/on_policy/second_order/pcpo.py [hunk 14]
```diff
@@ -119,31 +188,77 @@ def algorithm_specific_logs(self):
...
-        loss_pi, pi_info = self.compute_loss_pi(data=data)
+        # Process the advantage function.
+        processed_adv = self.compute_surrogate(adv=adv, cost_adv=cost_adv)
+        # Compute the loss of policy net.
+        loss_pi, pi_info = self.compute_loss_pi(obs=obs, act=act, log_p=log_p, adv=processed_adv)
         loss_pi_before = loss_pi.item()
         # get prob. distribution before updates
-        p_dist = self.actor_critic.actor(data['obs'])
+        p_dist = self.actor_critic.actor(obs)
         # Train policy with multiple steps of gradient descent
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/second_order/pcpo.py:257]: Lowercase

File: omnisafe/algorithms/on_policy/second_order/pcpo.py [hunk 15]
```diff
@@ -119,31 +188,77 @@ def algorithm_specific_logs(self):
...
+            adv (torch.Tensor): The advantage tensor.
+            cost_adv (torch.Tensor): The cost advantage tensor.
+        """
+        self.fvp_obs = obs[::4]
+        theta_old = get_flat_params_from(self.actor_critic.actor)
         self.actor_optimizer.zero_grad()
-        loss_pi, pi_info = self.compute_loss_pi(data=data)
+        # Process the advantage function.
+        processed_adv = self.compute_surrogate(adv=adv, cost_adv=cost_adv)
+        # Compute the loss of policy net.
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/second_order/pcpo.py:252]: Lowercase

File: omnisafe/algorithms/on_policy/second_order/pcpo.py [hunk 16]
```diff
@@ -119,31 +188,77 @@ def algorithm_specific_logs(self):
...
+            act (torch.Tensor): The action tensor.
+            log_p (torch.Tensor): The log probability of the action.
+            adv (torch.Tensor): The advantage tensor.
+            cost_adv (torch.Tensor): The cost advantage tensor.
+        """
+        self.fvp_obs = obs[::4]
+        theta_old = get_flat_params_from(self.actor_critic.actor)
         self.actor_optimizer.zero_grad()
-        loss_pi, pi_info = self.compute_loss_pi(data=data)
+        # Process the advantage function.
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/second_order/pcpo.py:250]: Lowercase

File: omnisafe/common/normalize.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+    def get_var(self):
+        """Get the variance."""
+        return self.var
+
+    def get_std(self):
+        """Get the std."""
+        return self.std
+
+    def forward(self, raw_data=None):
+        """FOrward"""
```
Comment (zmsn-2077) [inline omnisafe/common/normalize.py:73]: error spelling

File: omnisafe/wrappers/cmdp_wrapper.py [hunk 18]
```diff
@@ -0,0 +1,433 @@
...
+                        terminal_value, terminal_cost_value = 0.0, 0.0
+                    buf.finish_path(
+                        last_val=terminal_value,
+                        last_cost_val=terminal_cost_value,
+                        idx=idx,
+                    )
+                if terminal and not epoch_ended:
+                    # Log info about epoch
+                    self.rollout_log(logger, idx)
+                # Only save EpRet / EpLen if trajectory finished
```
Comment (zmsn-2077) [inline omnisafe/wrappers/cmdp_wrapper.py:341]: delete this line？

File: omnisafe/wrappers/cmdp_wrapper.py [hunk 19]
```diff
@@ -0,0 +1,433 @@
...
+            )
+            # Store values for statistic purpose
+            if self.rollout_data.use_cost:
+                logger.store(**{'Values/V': value, 'Values/C': cost_value})
+            else:
+                logger.store(**{'Values/V': value})
+            if use_rand_action:
+                action = self.env.action_space.sample()
+            # Step the env
+            # pylint: disable=unused-variable
```
Comment (zmsn-2077) [inline omnisafe/wrappers/cmdp_wrapper.py:388]: disable

File: omnisafe/wrappers/early_terminated_wrapper.py [hunk 20]
```diff
@@ -14,74 +14,56 @@
...
-                cost_val=cost_value,
-            )
 
-            # Store values for statistic purpose
-            if self.use_cost:
-                logger.store(**{'Values/V': value, 'Values/C': cost_value})
-            else:
-                logger.store(**{'Values/V': value})
+@WRAPPER_REGISTRY.register
+class EarlyTerminatedWrapper(CMDPWrapper):  # pylint: disable=too-many-instance-attributes
```
Comment (zmsn-2077) [inline omnisafe/wrappers/early_terminated_wrapper.py:30]: disable

File: omnisafe/algorithms/off_policy/td3_lag.py [hunk 21]
```diff
@@ -13,23 +13,27 @@
...
+import torch
 
 from omnisafe.algorithms import registry
 from omnisafe.algorithms.off_policy.td3 import TD3
 from omnisafe.common.lagrange import Lagrange
+from omnisafe.models.constraint_actor_q_critic import ConstraintActorQCritic
 
 
 @registry.register
 class TD3Lag(TD3, Lagrange):  # pylint: disable=too-many-instance-attributes
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/td3_lag.py:27]: disable

File: omnisafe/algorithms/off_policy/td3_lag.py [hunk 22]
```diff
@@ -13,23 +13,27 @@
...
     """The Lagrange version of the TD3 algorithm
```
Comment (zmsn-2077) [inline omnisafe/algorithms/off_policy/td3_lag.py:28]: .

File: omnisafe/algorithms/on_policy/pid_lagrange/trpo_pid.py [hunk 23]
```diff
@@ -41,44 +51,92 @@ def __init__(self, env_id, cfgs) -> None:
...
         self.logger.log_tabular('Metrics/LagrangeMultiplier', self.cost_penalty)
         self.logger.log_tabular('PID/pid_Kp', self.pid_kp)
         self.logger.log_tabular('PID/pid_Ki', self.pid_ki)
         self.logger.log_tabular('PID/pid_Kd', self.pid_kd)
 
-    def compute_loss_pi(self, data: dict):
-        """compute loss for policy"""
-        dist, _log_p = self.actor_critic.actor(data['obs'], data['act'])
-        ratio = torch.exp(_log_p - data['log_p'])
+    # pylint: disable=too-many-arguments
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/pid_lagrange/trpo_pid.py:76]: disable

File: omnisafe/algorithms/on_policy/base/ppo.py [hunk 24]
```diff
@@ -14,123 +14,76 @@
...
         self.clip = cfgs.clip
         super().__init__(
             env_id=env_id,
             cfgs=cfgs,
         )
 
-    def compute_loss_pi(self, data: dict):
-        """Compute policy loss."""
-        dist, _log_p = self.actor_critic.actor(data['obs'], data['act'])
+    # pylint: disable=too-many-arguments
```
Comment (zmsn-2077) [inline omnisafe/algorithms/on_policy/base/ppo.py:53]: disable

**Review by zmsn-2077 — CHANGES_REQUESTED** [review]
.

**Review by calico-1226 — APPROVED** [review]
I approve these changes.
